Sources/LanguageServerProtocol/SupportTypes/ServerCapabilities.swift

```diff
@@ -195,53 +231,60 @@ public enum ValueOrBool<ValueType: Codable>: Codable, Hashable where ValueType:
   }
 }
 
-public struct TextDocumentSyncOptions: Codable, Hashable {
-
-  /// Whether open/close notifications should be sent to the server.
-  public var openClose: Bool?
-
-  /// Whether and how the client should synchronize document changes with the server.
-  public var change: TextDocumentSyncKind?
-
-  /// Whether will-save notifications should be sent to the server.
-  public var willSave: Bool?
-
-  /// Whether will-save-wait-until notifications should be sent to the server.
-  public var willSaveWaitUntil: Bool?
+public enum PositionEncodingKind: String, Codable, Hashable {
+  case utf8 = "utf-8"
+  case utf16 = "utf-16"
+  case utf32 = "utf-32"
+}
 
-  public struct SaveOptions: Codable, Hashable {
+public enum TextDocumentSync: Codable, Hashable {
+  case options(TextDocumentSyncOptions)
+  case kind(TextDocumentSyncKind)
 
-    /// Whether the client should include the file content in save notifications.
-    public var includeText: Bool?
+  public init(from decoder: Decoder) throws {
+    if let options = try? TextDocumentSyncOptions(from: decoder) {
+      self = .options(options)
+    } else if let kind = try? TextDocumentSyncKind(from: decoder) {
+      self = .kind(kind)
+    } else {
+      let context = DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Expected TextDocumentSyncOptions or TextDocumentSyncKind")
+      throw DecodingError.dataCorrupted(context)
+    }
+  }
 
-    public init(includeText: Bool? = nil) {
-      self.includeText = includeText
+  public func encode(to encoder: Encoder) throws {
+    switch self {
+    case .options(let options):
+      try options.encode(to: encoder)
+    case .kind(let kind):
+      try kind.encode(to: encoder)
     }
   }
+}
 
-  /// Whether save notifications should be sent to the server.
-  public var save: ValueOrBool<SaveOptions>?
+public struct TextDocumentSyncOptions: Codable, Hashable {
+
+  /// Open and close notifications are sent to the server.
+  /// If omitted open close notifications should not be sent.
+  public var openClose: Bool?
+
+  /// Change notifications are sent to the server. See
+  /// TextDocumentSyncKind.None, TextDocumentSyncKind.Full and
+  /// TextDocumentSyncKind.Incremental. If omitted it defaults to
+  /// TextDocumentSyncKind.None.
+  public var change: TextDocumentSyncKind?
 
   public init(openClose: Bool? = true,
-              change: TextDocumentSyncKind? = .incremental,
-              willSave: Bool? = true,
-              willSaveWaitUntil: Bool? = false,
```

Comment:
I'm confused about what happened here. These fields were specified in LSP 3.14, but removed in 3.15 from the interface declaration.  But they're still mentioned on the request itself, e.g.

> Server Capability:
> 
> * property name (optional): textDocumentSync.willSaveWaitUntil
> * property type: boolean

From [LSP 3.17 - textDocument/willSaveWaitUntil](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_willSaveWaitUntil)

Reply:
It turns out the spec has two definitions of `TextDocumentSyncKind`:
- Without `change` and friends: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentSyncOptions
- With `change` and friends (scroll down a little from the link): https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didRename

Since VSCode still has the options (https://github.com/microsoft/vscode-languageserver-node/blob/f02ec883a68c6ee0e07a512c2114847f5f8c9d29/protocol/src/common/protocol.ts#L1697) let’s continue to include them.